### PR TITLE
Add JSON operators for serializing/deserializing files

### PR DESCRIPTION
**Pull request by kuujo**
This pull request adds new JSON operators for reading and writing JSON files and additional operators for encoding values to JSON:
* `JsonSerialize` serializes a tuple to a line-delimited JSON file
* `JsonDeserialize` deserializes a tuple from a line-delimited JSON file
* `ToJson` encodes a value to JSON
* `ToJsonObject` encodes a record to a JSON object
* `ToJsonArray` encodes a record to a JSON array

JSON parsing and encoding is done using the [Jackson](https://github.com/FasterXML/jackson) library. Dependencies are added to the build to download Jackson jars.

Tests are added for the `ToJson`, `ToJsonObject` and `ToJsonArray` operators, but I wasn't sure how to test the file-dependent operators.

### CODE REVIEW

**Comment by kuujo** [comment]
#8

File: modules/Json.tla [hunk 1]
```diff
@@ -8,29 +8,40 @@ LOCAL INSTANCE TLC
...
-     ELSE ", " \o ToJsonKeyValue(F, d) \o IF D \ {d} = {}
-                                          THEN "}" 
-                                          ELSE ToJson(F, D \ {d})
+ToJson(value) ==
+  (*************************************************************************)
+  (* Converts the given value to a JSON string. Records are converted to   *)
+  (* JSON objects, tuples to JSON arrays, and scalar values as their JSON  *)
+  (* representation.                                                       *)
+  (*************************************************************************)
+  ""
```
Comment (lemmy) [inline modules/Json.tla]: If would be nice to have (high-level) definitions of all operators to document what they do (no need to actual evaluate to the same value as the module override).  See e.g. definition of `JavaTime` at https://github.com/tlaplus/Examples/blob/536628fc9ec15263f17e5849f0a9a8e2fa20241d/specifications/SpecifyingSystems/TLC/TLC.tla#L8

Comment (kuujo) [reply modules/Json.tla]: Okay that makes sense

File: build.xml [hunk 2]
```diff
@@ -22,8 +22,20 @@
 		<get src="https://nightly.tlapl.us/dist/tla2tools.jar" dest="${lib}"/>
 	</target>
 
-	<target name="compile" depends="download" description="compile the java module overwrites">
-		<javac srcdir="${src}" destdir="${build}" classpath="${lib}/tla2tools.jar"
+	<target name="download-jackson-core" depends="init" description="downloads jackson-core-2.10.2.jar" unless="skip.download">
```
Comment (lemmy) [inline build.xml]: Consider adding a directory with `jackson-*.jar` to the git repo to be independent of maven.org.

File: tests/JsonTests.tla [hunk 3]
```diff
@@ -1,5 +1,15 @@
 ----------------------------- MODULE JsonTests -----------------------------
-EXTENDS Json, TLCExt
+EXTENDS Json, Naturals, TLCExt
+
+ASSUME(AssertEq(ToJson(<< >>), "[]"))
+ASSUME(AssertEq(ToJson(<<"a", "b", "c">>), "[\"a\",\"b\",\"c\"]"))
+ASSUME(AssertEq(ToJson([i \in 1..3 |-> "a"]), "{\"1\":\"a\",\"2\":\"a\",\"3\":\"a\"}"))
```
Comment (lemmy) [inline tests/JsonTests.tla]: Does it make sense to de-serialize a large/real-world .json file?

Comment (kuujo) [reply tests/JsonTests.tla]: Yeah, this is something left todo. I think the serialize/deserialize operators need tests.

**Comment by kuujo** [comment]
I'm trying to ensure this PR is consistent with the added types supported by #8

File: tests/JsonTests.tla (hunk 3 above)
Comment (lemmy) [reply tests/JsonTests.tla]: Maybe:

```tla
LET someTLA+Expr == ...
IN Deserialize(Serialize(someTLA+Expr, out.json), out.json) = someTLA+Expr
```

Comment (kuujo) [reply tests/JsonTests.tla]: I'll try it out

File: build.xml (hunk 2 above)
Comment (kuujo) [reply build.xml]: @lemmy where should I put it? Add a `lib` directory?

Comment (kuujo) [reply build.xml]: The `lib` directory is ignored

**Comment by kuujo** [comment]
@lemmy I think I got everything covered. The converters now support all the types in #8, there are tests for `JsonSerialize` and `JsonDeserialize`, and the Jackson jars are stored locally rather than downloaded.

I modified `.gitignore` to only ignore the `tla2tools.jar` in the `lib` directory and put the Jackson jars there. I'm not familiar with Ant so that may be a bad idea. I can put them somewhere else if you have any suggestions, but that seemed the most sensible.

**Comment by kuujo** [comment]
Ahh wait I didn't add specs for the operators...

**Comment by lemmy** [comment]
Thanks for your contribution!

Merged in https://github.com/tlaplus/CommunityModules/commit/1b78c803ce60a22e1692921fa5c70ba40da646d4

**Pull request closed without merging**